executor/cte_table_reader.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/cteutil"
+)
+
+// CTETableReaderExec scans data in iterInTbl, which is filled by corresponding CTEExec.
+type CTETableReaderExec struct {
+	baseExecutor
+
+	iterInTbl cteutil.Storage
+	chkIdx    int
+	curIter   int
+}
+
+// Open implements the Executor interface.
+func (e *CTETableReaderExec) Open(ctx context.Context) error {
+	e.reset()
+	return e.baseExecutor.Open(ctx)
+}
+
+// Next implements the Executor interface.
+func (e *CTETableReaderExec) Next(ctx context.Context, req *chunk.Chunk) (err error) {
+	req.Reset()
+
+	// Wait until iterInTbl can be read. This is controlled by corresponding CTEExec.
+	<-e.iterInTbl.GetBegCh()
+
+	// We should read `iterInTbl` from the beginning when the next iteration starts.
+	// Can not directly judge whether to start the next iteration based on e.chkIdx,
+	// because some operators(Selection) may use forloop to read all data in `iterInTbl`.
+	if e.curIter != e.iterInTbl.GetIter() {
```

Comment:
We do not need `GetIter` and `SetIter`, which we need actually is `IsNewIter`

Reply:
But we sitll need `SetNewIter` to set it to false. Guess there is no benefits to change?